virtual_ecosystem/models/soil/all_pools.py

```diff
@@ -64,145 +68,179 @@ class EnzymeMediatedRates:
-    delta_pools_ordered: dict[str, NDArray[np.float32]],
-    core_constants: CoreConsts,
-    model_constants: SoilConsts,
-) -> NDArray[np.float32]:
-    """Calculate net change for each carbon pool.
+class SoilPools:
+    """This class collects all the various soil pools so that they can be updated.
 
-    This function calls lower level functions which calculate the transfers between
-    pools. When all transfers have been calculated the net transfer is used to
-    calculate the net change for each pool.
+    This class contains a method to update all soil pools. As well as taking in the data
+    object it also has to take in another dictionary containing the pools. This
+    dictionary is modifiable by the integration algorithm whereas the data object will
+    only be modified when the entire soil model simulation has finished.
+    """
 
-    Args:
-        soil_c_pool_lmwc: Low molecular weight carbon pool [kg C m^-3]
-        soil_c_pool_maom: Mineral associated organic matter pool [kg C m^-3]
-        soil_c_pool_microbe: Microbial biomass (carbon) pool [kg C m^-3]
-        soil_c_pool_pom: Particulate organic matter pool [kg C m^-3]
-        soil_c_pool_necromass: Microbial necromass pool [kg C m^-3]
-        soil_enzyme_pom: Amount of enzyme class which breaks down particulate organic
-            matter [kg C m^-3]
-        soil_enzyme_maom: Amount of enzyme class which breaks down mineral associated
-            organic matter [kg C m^-3]
-        pH: pH values for each soil grid cell [unitless]
-        bulk_density: bulk density values for each soil grid cell [kg m^-3]
-        soil_moisture: amount of water contained by each soil layer [mm]
-        soil_water_potential: Soil water potential for each grid cell [kPa]
-        vertical_flow_rate: The vertical flow rate [TODO]
-        soil_temp: soil temperature for each soil grid cell [degrees C]
-        clay_fraction: The clay fraction for each soil grid cell [unitless]
-        mineralisation_rate: Amount of litter mineralised into POM pool [kg C m^-3
-            day^-1]
-        delta_pools_ordered: Dictionary to store pool changes in the order that pools
-            are stored in the initial condition vector.
-        core_constants: Set of constants shared between models.
-        model_constants: Set of constants for the soil model.
+    def __init__(
+        self, data: Data, pools: dict[str, NDArray[np.float32]], constants: SoilConsts
+    ):
+        self.data = data
+        """The data object for the Virtual Ecosystem simulation."""
+        self.pools = pools
+        """Pools which can change during the soil model update.
+        
+        These pools need to be added outside the data object otherwise the integrator
+        cannot update them and the integration will fail.
+        """
+        self.constants = constants
+
+    def calculate_all_pool_updates(
+        self,
+        delta_pools_ordered: dict[str, NDArray[np.float32]],
```

Comment:
This is because `scipy.solve_ivp` can only operate on a `numpy` array, but the data we are trying to update is stored as a dictionary. So something is needed to consistently relate the dictionary entries to specific rows in the`numpy` array. 

The solution I've used here is to create the dictionary that updated output is added to (`delta_pools_ordered`) in a higher level function so that the order of dictionary keys is always the same, which ensures that calling

`np.concatenate(list(delta_pools_ordered.values()))`

always results in a `numpy` array where a given dictionary key (e.g. "soil_n_pool_don") will always be used to fill in the same group of rows (e.g. rows 1-9).

There might be a smarter way to pass this kind of information about order around, but it didn't occur to me. If you have any suggestions for a cleaner way to approach this that would be useful!

Reply:
Ah, ok I get it. So the dictionary format - having keys and values - is a convenience thing. It is not really used except for ensuring the order of the data when transforming it into/from an array, right?

Well, it is pretty neat, to be honest. I cannot think on a better way of doing it. I would, however, explain this somewhere - maybe the docstring of the high level function - otherwise it is quite confusing and difficult to understand why it is the way it is.